cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/system/ArtifactSystemMetadataWriter.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright © 2016 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.data2.metadata.system;
+
+import co.cask.cdap.api.artifact.ArtifactClasses;
+import co.cask.cdap.api.plugin.PluginClass;
+import co.cask.cdap.data2.metadata.store.MetadataStore;
+import co.cask.cdap.proto.Id;
+import co.cask.cdap.proto.artifact.ArtifactInfo;
+import com.google.common.collect.ImmutableMap;
+
+import java.util.Map;
+import java.util.Set;
+import javax.annotation.Nullable;
+
+/**
+ * A {@link AbstractSystemMetadataWriter} for an {@link Id.Artifact artifact}.
+ */
+public class ArtifactSystemMetadataWriter extends AbstractSystemMetadataWriter {
+
+  private static final String PLUGIN_KEY_PREFIX = "plugin";
+
+  private final ArtifactInfo artifactInfo;
+
+  public ArtifactSystemMetadataWriter(MetadataStore metadataStore, Id.Artifact artifactId,
+                                      ArtifactInfo artifactInfo) {
+    super(metadataStore, artifactId);
+    this.artifactInfo = artifactInfo;
+  }
+
+  @Override
+  Map<String, String> getSystemPropertiesToAdd() {
+    ImmutableMap.Builder<String, String> properties = ImmutableMap.builder();
+    ArtifactClasses classes = artifactInfo.getClasses();
+    for (PluginClass pluginClass : classes.getPlugins()) {
+      String name = pluginClass.getName();
+      String type = pluginClass.getType();
+      // Need both name and type in the key because two plugins of different types could have the same name.
+      // However, the composite of name + type is guaranteed to be unique
+      properties.put(PLUGIN_KEY_PREFIX + COLON + name + COLON + type, name + COLON + type);
```

Comment:
The following queries should work:
1. ``<plugin name>``
2. ``<plugin type>``
3. ``<plugin name>:<plugin type>``
4. ``plugin:<plugin name>:<plugin type>``.

However, I'm probably missing a test for this, will add.